File: func_parser/scheduler.py
"""Task scheduler for func_parser."""
from __future__ import annotations

import asyncio
import re
from datetime import datetime, timedelta
from typing import Callable, List, Optional

from .core.errors import SchedulerError

__all__ = ["ScheduledTask", "Scheduler"]


class ScheduledTask:
    """Represents a scheduled command execution."""

    def __init__(self, spec: str, command: str, callback: Callable) -> None:
        self.spec = spec
        self.command = command
        self.callback = callback
        self._task: Optional[asyncio.Task] = None

    def cancel(self) -> None:
        """Cancel this scheduled task."""
        if self._task:
            self._task.cancel()


class Scheduler:
    """Simple task scheduler supporting ``every Xm/s/h`` and ``at HH:MM`` syntax."""

    INTERVAL_RE = re.compile(
        r'every\s+(\d+(?:\.\d+)?)\s*(s|sec|seconds?|m|min|minutes?|h|hrs?|hours?)',
        re.IGNORECASE,
    )
    AT_RE = re.compile(r'at\s+(\d{1,2}):(\d{2})', re.IGNORECASE)

    def __init__(self, execute_fn: Callable) -> None:
        """*execute_fn* is called with the command string at the scheduled time."""
        self._execute = execute_fn
        self._tasks: List[ScheduledTask] = []

    def parse_spec(self, spec: str) -> float:
        """Parse a schedule spec and return seconds until the next execution.

        Supports:
        - ``every 5m`` / ``every 30s`` / ``every 2h``
        - ``at 12:00``
        """
        m = self.INTERVAL_RE.match(spec.strip())
        if m:
            amount = float(m.group(1))
            unit = m.group(2).lower()
            if unit.startswith("s"):
                return amount
            if unit.startswith("m"):
                return amount * 60
            if unit.startswith("h"):
                return amount * 3600

        m = self.AT_RE.match(spec.strip())
        if m:
            now = datetime.now()
            target = now.replace(
                hour=int(m.group(1)),
                minute=int(m.group(2)),
                second=0,
                microsecond=0,
            )
            if target <= now:
                target += timedelta(days=1)
            return (target - now).total_seconds()

        raise SchedulerError(f"Cannot parse schedule spec: {spec!r}")

    def schedule(self, spec: str, command: str) -> ScheduledTask:
        """Schedule *command* according to *spec*.  Returns a :class:`ScheduledTask`."""
        is_repeating = bool(self.INTERVAL_RE.match(spec.strip()))

        async def _run() -> None:
            if is_repeating:
                while True:
                    delay = self.parse_spec(spec)
                    await asyncio.sleep(delay)
                    await self._execute(command)
            else:
                delay = self.parse_spec(spec)
                await asyncio.sleep(delay)
                await self._execute(command)

        task_obj = ScheduledTask(spec, command, self._execute)
        task_obj._task = asyncio.ensure_future(_run())
        self._tasks.append(task_obj)
        return task_obj

    def cancel_all(self) -> None:
        """Cancel all scheduled tasks."""
        for t in self._tasks:
            t.cancel()
        self._tasks.clear()

File: func_parser_rs/src/scheduler.rs
//! Task scheduler for func_parser_rs.

use std::time::Duration;
use crate::errors::{FuncParserError, Result};

/// Parse a schedule spec and return the initial delay in seconds.
///
/// Supported formats:
/// - `every 5m` / `every 30s` / `every 2h`
/// - `at 12:00`
pub fn parse_spec(spec: &str) -> Result<f64> {
    let s = spec.trim();

    // every <N><unit>
    if let Some(rest) = s.strip_prefix("every ") {
        let rest = rest.trim();
        // split at the transition from digit to alpha
        let split = rest
            .find(|c: char| c.is_alphabetic())
            .ok_or_else(|| FuncParserError::SchedulerError(format!("cannot parse: {:?}", spec)))?;
        let amount: f64 = rest[..split].trim().parse().map_err(|_| {
            FuncParserError::SchedulerError(format!("invalid number in: {:?}", spec))
        })?;
        let unit = rest[split..].trim().to_lowercase();
        let secs = if unit.starts_with('s') {
            amount
        } else if unit.starts_with('m') {
            amount * 60.0
        } else if unit.starts_with('h') {
            amount * 3_600.0
        } else {
            return Err(FuncParserError::SchedulerError(format!("unknown unit in: {:?}", spec)));
        };
        return Ok(secs);
    }

    // at HH:MM
    if let Some(rest) = s.strip_prefix("at ") {
        let rest = rest.trim();
        let parts: Vec<&str> = rest.split(':').collect();
        if parts.len() != 2 {
            return Err(FuncParserError::SchedulerError(format!("invalid time in: {:?}", spec)));
        }
        let hour: u32 = parts[0].trim().parse().map_err(|_| {
            FuncParserError::SchedulerError(format!("invalid hour in: {:?}", spec))
        })?;
        let minute: u32 = parts[1].trim().parse().map_err(|_| {
            FuncParserError::SchedulerError(format!("invalid minute in: {:?}", spec))
        })?;
        if hour > 23 || minute > 59 {
            return Err(FuncParserError::SchedulerError(format!(
                "time out of range in: {:?} (hour 0-23, minute 0-59)", spec
            )));
        }
        // Calculate seconds until next occurrence using wall-clock
        let now = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_secs();
        let secs_today = (now % 86_400) as u32; // approximate UTC seconds of day
        let target_secs = hour * 3_600 + minute * 60;
        let delay = if target_secs > secs_today {
            (target_secs - secs_today) as f64
        } else {
            (86_400 - secs_today + target_secs) as f64
        };
        return Ok(delay);
    }

    Err(FuncParserError::SchedulerError(format!("cannot parse schedule spec: {:?}", spec)))
}

/// A running scheduled task handle.
pub struct ScheduledTask {
    pub spec: String,
    pub command: String,
    pub handle: tokio::task::JoinHandle<()>,
}

impl ScheduledTask {
    pub fn abort(&self) {
        self.handle.abort();
    }
}

/// Simple task scheduler.
pub struct Scheduler {
    tasks: Vec<ScheduledTask>,
}

impl Scheduler {
    pub fn new() -> Self {
        Self { tasks: Vec::new() }
    }

    /// Schedule `command` according to `spec`.
    ///
    /// `execute_fn` is called with the command string at the scheduled time.
    pub fn schedule<F, Fut>(
        &mut self,
        spec: impl Into<String>,
        command: impl Into<String>,
        execute_fn: F,
    ) -> Result<()>
    where
        F: Fn(String) -> Fut + Send + 'static,
        Fut: std::future::Future<Output = ()> + Send,
    {
        let spec_str: String = spec.into();
        let cmd: String = command.into();
        let delay = parse_spec(&spec_str)?;

        let is_repeating = spec_str.trim().starts_with("every ");
        let spec_clone = spec_str.clone();
        let cmd_clone = cmd.clone();

        let handle = tokio::spawn(async move {
            if is_repeating {
                loop {
                    let d = parse_spec(&spec_clone).unwrap_or(60.0);
                    tokio::time::sleep(Duration::from_secs_f64(d)).await;
                    execute_fn(cmd_clone.clone()).await;
                }
            } else {
                tokio::time::sleep(Duration::from_secs_f64(delay)).await;
                execute_fn(cmd_clone).await;
            }
        });

        self.tasks.push(ScheduledTask { spec: spec_str, command: cmd, handle });
        Ok(())
    }

    pub fn cancel_all(&mut self) {
        for task in self.tasks.drain(..) {
            task.abort();
        }
    }
}

impl Default for Scheduler {
    fn default() -> Self {
        Self::new()
    }
}
